src/symmetric_memory.cpp

```diff
@@ -0,0 +1,98 @@
+/*
+ * #%L
+ * %%
+ * Copyright (C) 2020 BMW Car IT GmbH
+ * %%
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * #L%
+ */
+
+#include "mococrw/symmetric_memory.h"
+
+#include <queue>
+#include <cassert>
+
+namespace mococrw
+{
+void QueueOfVectorsMemoryStrategy::write(const std::vector<uint8_t> &&chunk)
+{
+    _totalBytesStored += chunk.size();
+
+    _queue.emplace_back(chunk);
+}
+
+std::vector<uint8_t> QueueOfVectorsMemoryStrategy::read(size_t requestedBufferSize)
+{
+    if (_queue.empty()) {
+        return std::vector<uint8_t>();
+    }
+
+    if (_queue.front().size() == requestedBufferSize) {
+        return _readCompleteBlock();
+    } else if (_queue.front().size() > requestedBufferSize) {
+        return _readPartialBlock(requestedBufferSize);
+    } else {
+        std::vector<uint8_t> bufferToReturn;
+        bufferToReturn.reserve(requestedBufferSize);
+
+        // First, read as much complete blocks as we can
+        while (!_queue.empty() &&
+               (_queue.front().size() <= (requestedBufferSize - bufferToReturn.size()))) {
+            auto block = _readCompleteBlock();
+            std::copy(std::begin(block), std::end(block), std::back_inserter(bufferToReturn));
+        }
+        // Then, read partial block if needed
+        if (bufferToReturn.size() < requestedBufferSize && !_queue.empty()) {
+            //size_t blockToRead = std::min(_queue.front().size(), requestedBufferSize - bufferToReturn.size());
+            size_t blockToRead = requestedBufferSize - bufferToReturn.size();
+            auto block = _readPartialBlock(blockToRead);
+            std::copy(std::begin(block), std::end(block), std::back_inserter(bufferToReturn));
+        }
+
+        return bufferToReturn;
+    }
+}
+
+std::vector<uint8_t> QueueOfVectorsMemoryStrategy::readAll()
+{
+    std::vector<uint8_t> bufferToReturn;
+    bufferToReturn.reserve(_totalBytesStored);
+    while (!_queue.empty()) {
+        std::copy(std::begin(_queue.front()), std::end(_queue.front()),
+                  std::back_inserter(bufferToReturn));
+        _queue.pop_front();
+    }
```

Comment:
I disagree. This is useful if there is only a single entry in the queue, and this will actually be a common case in a `update()/read()` or `update()/readAll()` loop. I've added some code to cover this case.